protocol-units/bridge/chains/ethereum/src/lib.rs

```diff
@@ -0,0 +1,199 @@
+use anyhow::Context;
+use std::fmt::Debug;
+
+use alloy::signers::local::PrivateKeySigner;
+use alloy_network::EthereumWallet;
+use alloy_primitives::private::serde::{Deserialize, Serialize};
+use alloy_primitives::{FixedBytes, U256};
+use alloy_provider::{Provider, ProviderBuilder};
+use alloy_rlp::{Decodable, RlpDecodable, RlpEncodable};
+use alloy_sol_types::sol;
+use bridge_shared::bridge_contracts::{
+	BridgeContractInitiator, BridgeContractInitiatorError, BridgeContractInitiatorResult,
+};
+use bridge_shared::types::{
+	Amount, BridgeTransferDetails, BridgeTransferId, HashLock, HashLockPreImage, InitiatorAddress,
+	RecipientAddress, TimeLock,
+};
+
+pub mod types;
+pub mod utils;
+
+use crate::types::{EthAddress, EthHash};
+
+const INITIATOR_CONTRACT: &str = "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266";
+const DEFAULT_GAS_LIMIT: u64 = 10_000_000_000;
+
+///Configuration for the Ethereum Bridge Client
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Config {
+	pub rpc_url: Option<String>,
+	pub ws_url: Option<String>,
+	pub chain_id: String,
+	pub signer_private_key: String,
+	pub initiator_contract: EthAddress,
+	pub gas_limit: u64,
+}
+
+impl Default for Config {
+	fn default() -> Self {
+		Config {
+			rpc_url: Some("http://localhost:8545".to_string()),
+			ws_url: Some("ws://localhost:8545".to_string()),
+			chain_id: "31337".to_string(),
+			signer_private_key: Self::default_for_private_key(),
+			initiator_contract: EthAddress::from(INITIATOR_CONTRACT.to_string()),
+			gas_limit: DEFAULT_GAS_LIMIT,
+		}
+	}
+}
+
+impl Config {
+	fn default_for_private_key() -> String {
+		let random_wallet = PrivateKeySigner::random();
+		random_wallet.to_bytes().to_string()
+	}
+}
+
+// Codegen from the abi
+sol!(
+	#[allow(missing_docs)]
+	#[sol(rpc)]
+	AtomicBridgeInitiator,
+	"abis/AtomicBridgeInitiator.json"
+);
+
+#[derive(RlpDecodable, RlpEncodable)]
+struct EthBridgeTransferDetails {
+	pub amount: U256,
+	pub originator: EthAddress,
+	pub recipient: [u8; 32],
+	pub hash_lock: [u8; 32],
+	pub time_lock: U256,
+	pub state: u8, // Assuming the enum is u8 for now..
+}
+
+pub struct EthClient<P> {
+	rpc_provider: P,
+	initiator_contract: EthAddress,
+}
+
+impl EthClient<utils::AlloyProvider> {
+	pub async fn build_with_config(config: Config) -> Result<Self, anyhow::Error> {
+		let signer = config.signer_private_key.parse::<PrivateKeySigner>()?;
+		let rpc_url = config.rpc_url.context("rpc_url not set")?;
+		let rpc_provider = ProviderBuilder::new()
+			.with_recommended_fillers()
+			.wallet(EthereumWallet::from(signer.clone()))
+			.on_builtin(&rpc_url)
+			.await?;
+
+		Ok(EthClient { rpc_provider, initiator_contract: config.initiator_contract })
+	}
+}
+
+impl<P> Clone for EthClient<P> {
+	fn clone(&self) -> Self {
+		todo!()
```

Comment:
No need to implement it, otherwise the abstract bridge will not function.

Reply:
See issue (with its own PR) https://github.com/movementlabsxyz/movement/issues/250